presto-native-execution/presto_cpp/main/tests/PeriodicMemoryCheckerTest.cpp

```diff
@@ -0,0 +1,216 @@
+        PeriodicMemoryChecker::Config config,
+        int64_t systemUsedMemoryBytes = 0,
+        int64_t mallocBytes = 0,
+        std::function<void()>&& periodicCb = nullptr,
+        std::function<bool(const std::string&)>&& heapDumpCb = nullptr)
+        : PeriodicMemoryChecker(config),
+          systemUsedMemoryBytes_(systemUsedMemoryBytes),
+          mallocBytes_(mallocBytes),
+          periodicCb_(std::move(periodicCb)),
+          heapDumpCb_(std::move(heapDumpCb)) {}
+
+    void setMallocBytes(int64_t mallocBytes) {
+      mallocBytes_ = mallocBytes;
+    }
+
+   protected:
+    int64_t systemUsedMemoryBytes() const override {
+      return systemUsedMemoryBytes_;
+    }
+
+    int64_t mallocBytes() const override {
+      return mallocBytes_;
+    }
+
+    void periodicCb() const override {
+      if (periodicCb_) {
+        periodicCb_();
+      }
+    }
+
+    bool heapDumpCb(const std::string& filePath) const {
+      if (heapDumpCb_) {
+        return heapDumpCb_(filePath);
+      }
+      return false;
+    }
+
+   private:
+    int64_t systemUsedMemoryBytes_{0};
+    int64_t mallocBytes_{0};
+    std::function<void()> periodicCb_;
+    std::function<bool(const std::string&)> heapDumpCb_;
+  };
+};
+
+TEST_F(PeriodicMemoryCheckerTest, basic) {
+  // Default config
+  ASSERT_NO_THROW(TestPeriodicMemoryChecker(PeriodicMemoryChecker::Config{}));
+
+  ASSERT_NO_THROW(TestPeriodicMemoryChecker(PeriodicMemoryChecker::Config{
+      1, true, 1024, 32, true, 5, "/path/to/dir", "prefix", 5, 512}));
+  ASSERT_THROW(
+      TestPeriodicMemoryChecker(PeriodicMemoryChecker::Config{
+          1, true, 0, 32, true, 5, "/path/to/dir", "prefix", 5, 512}),
+      VeloxException);
+  ASSERT_THROW(
+      TestPeriodicMemoryChecker(PeriodicMemoryChecker::Config{
+          1, true, 1024, 32, true, 5, "", "prefix", 5, 512}),
+      VeloxException);
+  ASSERT_THROW(
+      TestPeriodicMemoryChecker(PeriodicMemoryChecker::Config{
+          1, true, 1024, 32, true, 5, "/path/to/dir", "", 5, 512}),
+      VeloxException);
+  TestPeriodicMemoryChecker memChecker(PeriodicMemoryChecker::Config{
+      1, false, 0, 0, false, 5, "/path/to/dir", "prefix", 5, 512});
+  ASSERT_NO_THROW(memChecker.start());
+  ASSERT_THROW(memChecker.start(), VeloxException);
+  ASSERT_NO_THROW(memChecker.stop());
+}
+
+TEST_F(PeriodicMemoryCheckerTest, periodicCb) {
+  auto testPeriodicCb = [](bool pushbackEnabled, bool heapDumpEnabled) {
+    std::atomic_bool periodicCbInvoked{false};
+    TestPeriodicMemoryChecker memChecker(
+        PeriodicMemoryChecker::Config{
+            1,
+            pushbackEnabled,
+            512,
+            32,
+            heapDumpEnabled,
+            5,
+            "/path/to/dir",
+            "prefix",
+            5,
+            512},
+        pushbackEnabled ? 768 : 256,
+        128,
+        [&]() { periodicCbInvoked = true; });
+    memChecker.start();
+    std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
i wonder if this makes the test flaky, since 1 second is not strictly guaranteed as it can depend on when the thread gets to run after it receives the wake for a 1 sec interval. Do we support 0 seconds to ensure FunctionScheduler never sleeps?

Reply:
Thanks @bikramSingh91 , FunctionScheduler will immediately schedule the job at 0th second by default, if no delay is specified.